### PR TITLE
Prevent errors when using the caret plugin on an empty jQuery collection.

### PR DESCRIPTION
**Pull request by obecker**
If a jQuery collection has no elements the plugin should behave like other get/set functions (for example val() or prop()), i.e. the setter should do nothing and return this, the getter should return a default value (I think it should return undefined, with this change it returns 0).

### CODE REVIEW

File: jquery.caret.js [hunk 1]
```diff
@@ -1,79 +1,84 @@
 (function($) {
   $.fn.caret = function(pos) {
     var target = this[0];
-	var isContentEditable = target.contentEditable === 'true';
+    var isContentEditable = target && target.contentEditable === 'true';
+    var range, range1, range2;
     //get
-    if (arguments.length == 0) {
+    if (arguments.length === 0) {
+      if (target) {
```
Comment (accursoft) [inline jquery.caret.js]: This should probably just be `if (target) return`, instead of adding a new level of nesting.

File: jquery.caret.js [hunk 2]
```diff
@@ -1,79 +1,84 @@
...
+          range.moveToBookmark(bookmark);
+          while (range.moveStart('character', -1) !== 0) pos++;
+          return pos;
+        }
+        // Addition for jsdom support
+        if (target.selectionStart)
+          return target.selectionStart;
+      }
+      //not supported
+      return 0;
```
Comment (accursoft) [inline jquery.caret.js]: and this should also change to return.

File: jquery.caret.js [hunk 3]
```diff
@@ -1,79 +1,84 @@
...
+    }
+    //set
+    if (target) {
```
Comment (accursoft) [inline jquery.caret.js]: Likewise `if (!target) return this`.

File: jquery.caret.js [hunk 4]
```diff
@@ -1,79 +1,84 @@
 (function($) {
   $.fn.caret = function(pos) {
     var target = this[0];
```
Comment (accursoft) [inline jquery.caret.js]: Alternatively, maybe just abandon here if `!target`?

File: jquery.caret.js [hunk 5]
```diff
@@ -1,79 +1,84 @@
 (function($) {
   $.fn.caret = function(pos) {
     var target = this[0];
-	var isContentEditable = target.contentEditable === 'true';
+    var isContentEditable = target && target.contentEditable === 'true';
```
Comment (accursoft) [inline jquery.caret.js]: Alternatively, maybe just abandon here if `!target`?

File: jquery.caret.js (hunk 1 above)
Comment (obecker) [inline jquery.caret.js]: Actually I had this in my first version, however, I prefer to have one (or at least as few as possible) return points in a function. Especially here when a group of return statements should return the same value. Feel free to change it the way you like.

File: jquery.caret.js (hunk 5 above)
Comment (obecker) [inline jquery.caret.js]: and then `return arguments.length === 0 ? 0 : this`?
As I said below, I would prefer fewer return statements and no repeated expressions (the comparison with `arguments.length`) even if this means a deeper nesting. I don't want to argue about code style. Do as you like.

**Comment by accursoft** [comment]
OK, I've added an abandon ship for `!target`. It doesn't return `this` when setting a null target, which might be a good thing i.e. there's some indication to the caller that there was an error.

Let me know if this causes any problems for your use case.

**Comment by accursoft** [comment]
As for collecting the variables, this will need rebasing. Maybe rename `range` to `range1` while you're at it, does this actually need 3 variables for ranges?

**Comment by obecker** [comment]
I think chaining should work, even if the jquery object to be modified is empty.
For example

    $('...').caret(1).css('background-color', 'yellow');

must never raise an exception. So returning `undefined` when setting a value is indeed a problem.

**Comment by accursoft** [comment]
OK, I'll accept your original commit but might be a while until I get a chance to see to it.

**Comment by accursoft** [comment]
Implemented in 8cd0910544f4267d91ca527117ae459832978411, couldn't merge it because of other changes so I just re-implemented it.

**Pull request closed without merging**